grove/connectors/gsuite/usage.py

```diff
@@ -0,0 +1,230 @@
+from grove.constants import CHRONOLOGICAL
+from grove.exceptions import (
+    ConfigurationException,
+    NotFoundException,
+    RequestFailedException,
+)
+
+# This connector is only interested in GSuite Usage Reports.
+SCOPES = ["https://www.googleapis.com/auth/admin.reports.usage.readonly"]
+
+
+def as_rfc3339(date: datetime) -> str:
+    """Formats an input date into RFC3339 format.
+
+    :param date: The input datetime object to convert to RFC3339 format.
+
+    :return: An RFC 3339 format date as a string.
+    """
+    return date.isoformat(sep="T", timespec="milliseconds") + "Z"
+
+
+class Connector(BaseConnector):
+    CONNECTOR = "gsuite_usage"
+    POINTER_PATH = "date"
+    LOG_ORDER = CHRONOLOGICAL
+
+    @property
+    def delay(self):
+        """Defines the amount of time to delay collection of logs (in minutes).
+
+        This is used to allow time for logs to become 'consistent' before they are
+        collected. Google backfill log entries based on their published lag guidelines.
+        As a result of this, collection of events within this lag window may result in
+        missed events.
+
+        As a result of these constraints, this value is configurable to allow operators
+        to preference consistency over speed of delivery, and vice versa. For example,
+        a delay of 20 would instruct Grove to only collect logs after they are at least
+        20 minutes old.
+
+        This defaults to 0 (no delay).
+
+        :return: The "delay" component of the connector configuration.
+        """
+        try:
+            candidate = self.configuration.delay
+        except AttributeError:
+            return 0
+
+        try:
+            candidate = int(candidate)
+        except ValueError as err:
+            raise ConfigurationException(
+                f"Configured 'delay' is not valid. Value must be an integer. {err}"
+            )
+
+        return candidate
+
+    def collect(self):
+        """Collects usage reports from the Google GSuite Reports API.
+
+        This method retrieves usage reports for the entire enterprise or users for a specific date range,
+        starting from 7 days ago if no pointer is stored, and paginates through the results.
+
+        :raises RequestFailedException: An HTTP request failed.
+        """
+        self.logger.debug("Beginning data collection from GSuite API.")
+        cursor = None
+        http = google_auth_httplib2.AuthorizedHttp(
+            self.get_credentials(),
+            http=self.get_http_transport(),
+        )
+
+        # Determine the report type to query.
+        try:
+            report_type = self.configuration.usage_report_type
+            if report_type not in ["customerUsageReports", "userUsageReports", "entityUsageReports"]:
+                raise ConfigurationException(
+                    f"Invalid usage_report_type: {report_type}. Must be 'customerUsageReports', 'userUsageReports', or 'entityUsageReports'."
+                )
```

Comment:
Minor: Add vertical space, and remove duplication of report types.

```suggestion
            report_type = self.configuration.usage_report_type
            supported_types = ["customerUsageReports", "userUsageReports", "entityUsageReports"]

            if report_type not in supported_types:
                raise ConfigurationException(
                    f"Invalid usage_report_type: {report_type}. Must be one of {supported_types}"
                )
```

Reply:
[cc5d449](https://github.com/hashicorp-forge/grove/pull/84/commits/cc5d449d02ada473a8ce7d40f7dc9e76b7a2c704)